### PR TITLE
Remove Row/Column feature

### PR DESCRIPTION
**Pull request by dmitrii-kuznetsov**
Added remove row/column feature and changed how row/column adding works to make it less glitchy and more predictable.
While initially based on Toolbars were removed and `renderSettings` API has been used instead.

There are six new buttons in settings toolbar - all of them are self-explanatory (I hope):
- Add column after current
- Add column before current
- Add row after current
- Add row before current
- Remove current column
- Remove current row

### CODE REVIEW

**Comment by neSpecc** [comment]
Wow, cool

**Comment by DennieMello** [comment]
> Added remove row/column feature and changed how row/column adding works to make it less glitchy and more predictable.
> While initially based on Toolbars were removed and `renderSettings` API has been used instead.
> 
> There are six new buttons in settings toolbar - all of them are self-explanatory (I hope):
> 
> * Add column after current
> * Add column before current
> * Add row after current
> * Add row before current
> * Remove current column
> * Remove current row

Дмитрий, спасибо большое. Очень нахватало этого функционала! Возможно в будущем вы сможете сделать еще одну необходимую вещь - возможность выбрать заголовочные ячейки таблицы (сверху, слева и там и там) - http://htmlbook.ru/html/th.

**Comment by JimishF** [comment]
Hello? Review this please.

**Comment by ranemihir** [comment]
@neSpecc 
I forked the PR and tested it.
I actually have a different opinion on this.

The PR provides two buttons for deleting the selected row and selected column in the block settings (i.e. renderSettings). After deleting the selected row/column, if the delete row/delete column button is pressed again the last or the most recent row/column gets deleted. The last rows/columns keep deleting if keep pressing on the delete row/delete column button.

But what if say, a table has 100 rows and one wants to delete 20 rows in the middle somewhere. Then, he/she has to select the row/column and then press the delete row/delete column button for each and every row. I think that's a significant issue.

This can be done quite easily in the old UI i.e. PR #48 , by just keep on pressing the red minus button. Also, what I feel is the UI used in PR #48 is much more intuitive than this UI, even though it needs a lot work.

I do like the fact that the selected cell gets highlighted on focus.

File: src/plugin.js [hunk 1]
```diff
@@ -40,10 +56,78 @@ class Table {
...
    */
-  constructor({data, config, api}) {
+  constructor({ data, config, api }) {
     this.api = api;
 
     this._tableConstructor = new TableConstructor(data, config, api);
+
+    this.actions = [
+      {
+        actionName: 'InsertColBefore',
```
Comment (neSpecc) [inline src/plugin.js:66]: Please, add human-readable labels for displaying them via [Tooltip API](https://editorjs.io/tooltip-api). Also, all texts should be wrapped with [I18n API](https://editorjs.io/i18n-api)

[Example](https://github.com/editor-js/image/blob/master/src/tunes.js#L87)

File: src/plugin.js [hunk 2]
```diff
@@ -40,10 +56,78 @@ class Table {
...
+        actionName: 'DeleteCol',
+        icon: Icons.DeleteCol
+      }
+    ];
+  }
+
+  /**
+   * perform selected action
+   * @param actionName {string} - action name
+   */
```
Comment (neSpecc) [inline src/plugin.js:95]: please, describe a type of returning value.

File: src/plugin.js [hunk 3]
```diff
@@ -40,10 +56,78 @@ class Table {
...
+   * render actions toolbar
+   * @returns {HTMLDivElement}
+   */
+  renderSettings() {
+    const wrapper = document.createElement('div');
+
+    this.actions.forEach(({ actionName, icon }) => {
+      const button = document.createElement('div');
+
+      button.classList.add('ce-settings__button');
```
Comment (neSpecc) [inline src/plugin.js:123]: please, use the class name from the [Styles API](https://editorjs.io/styles#settingsbutton)

File: src/styles/table.pcss [hunk 4]
```diff
@@ -34,6 +34,10 @@
     overflow: hidden;
   }
 
+  &__highlight {
+    background-color: rgba(0, 0, 0, .1);
```
Comment (neSpecc) [inline src/styles/table.pcss:38]:
```suggestion
    background-color: rgba(173, 164, 176, 0.1);
```

File: src/table.js [hunk 5]
```diff
@@ -7,6 +7,7 @@ const CSS = {
   cell: 'tc-table__cell',
   wrapper: 'tc-table__wrap',
   area: 'tc-table__area',
```
Comment (neSpecc) [inline src/table.js:9]: do we still need this element? It would be great to remove it if it is not used anymore.

File: src/tableConstructor.js [hunk 6]
```diff
@@ -102,310 +94,15 @@ export class TableConstructor {
...
-    if (!(this._table.selectedCell !== null && !event.shiftKey)) {
-      return;
-    }
-    const indicativeRow = this._table.selectedCell.closest('TR');
-    let index = this._getHoveredSideOfContainer();
-
-    if (index === 1) {
-      index = indicativeRow.sectionRowIndex + 1;
-    }
-    const newstr = this._table.addRow(index);
```
Comment (neSpecc) [inline src/tableConstructor.js:389]: can we save the feature of adding of row by the Enter keydown?

**Comment by neSpecc** [comment]
> But what if say, a table has 100 rows and one wants to delete say 60 rows in the middle somewhere. Then, he/she has to select the row/column and then press the delete row/delete column button for each and every row. I think that's a significant issue.

It would be great if there will be an ability to select several lines by Shift (or Cmd) + arrows, or with the mouse. Then they can be deleted by Backspace/Delete/Enter button or by Remove Row button from the Block Tunes.

I think it can be implemented via separate PR.

@ranemihir  @dmitrii-kuznetsov

File: src/table.js (hunk 5 above)
Comment (dmitrii-kuznetsov) [reply src/table.js:9]: Yes, we do. [This element is still used](https://github.com/editor-js/table/pull/51/files/038330435804dc993bab8d5a23f26be5ed7adb67#diff-7612ce4df087606e51e938e7feb1ebceL111)

File: src/tableConstructor.js (hunk 6 above)
Comment (dmitrii-kuznetsov) [reply src/tableConstructor.js:389]: Sure. What combination should be used? You say just `Enter` but judging by the code it combination is `Shift+Enter` and comment to this function says the combination is `Ctrl+Enter` 👀

File: src/styles/table.pcss (hunk 4 above)
Comment (dmitrii-kuznetsov) [reply src/styles/table.pcss]: Changed.

File: src/plugin.js (hunk 3 above)
Comment (dmitrii-kuznetsov) [reply src/plugin.js]: Changed

File: src/plugin.js (hunk 2 above)
Comment (dmitrii-kuznetsov) [reply src/plugin.js:102]: Updated

File: src/plugin.js (hunk 1 above)
Comment (dmitrii-kuznetsov) [reply src/plugin.js:66]: Changed

**Comment by JimishF** [comment]
It seems that undo doesn't work on table.

**Comment by ranemihir** [comment]
@neSpecc Shift + arrows is currently being used for selecting blocks by default. Some other command will have to be used.

**Comment by robonetphy** [comment]
Hi @dmitrii-kuznetsov,
There are some errors in your PR, You must check the falling condition of your code I have found some which are as follows:
- Add row not working.
    - Add Table Element Delete all rows and click on delete row twice then you can't add any rows or columns.
    - I analyse the behaviour and came to the conclusion that if you press delete rows or columns on an empty table that is causing some problem in your logic
- Please add the Title to each tune, its useful for users.

**Comment by Davidzhu001** [comment]
@dmitrii-kuznetsov are you still maintaini ng?

**Pull request closed without merging**